src/lib.rs

```diff
@@ -332,31 +381,396 @@ impl<F: PrimeField, PC: PolynomialCommitment<F, DensePolynomial<F>>, FS: FiatSha
+        // correct degree.
+        let index_info = index_vk.index_info;
+        let degree_bounds = vec![None; index_vk.index_comms.len()]
+            .into_iter()
+            .chain(AHPForR1CS::prover_first_round_degree_bounds(&index_info))
+            .chain(AHPForR1CS::prover_second_round_degree_bounds(&index_info))
+            .chain(AHPForR1CS::prover_third_round_degree_bounds(&index_info))
+            .collect::<Vec<_>>();
+
+        // Gather commitments in one vector.
+        let commitments: Vec<_> = index_vk
+            .iter()
+            .chain(first_comms)
+            .chain(second_comms)
+            .chain(third_comms)
+            .cloned()
+            .zip(AHPForR1CS::<F>::polynomial_labels())
+            .zip(degree_bounds)
+            .map(|((c, l), d)| LabeledCommitment::new(l, c, d))
+            .collect();
+
+        let (query_set, verifier_state) =
+            AHPForR1CS::verifier_query_set(verifier_state, &mut fs_rng);
+
+        fs_rng.absorb(&to_bytes(&proof.evaluations));
+        let mut opening_challenge: ChallengeGenerator<F, S> =
+            ChallengeGenerator::new_multivariate(fs_rng);
+
+        let mut evaluations = Evaluations::new();
+        let mut evaluation_labels = Vec::new();
+        for (poly_label, (_, point)) in query_set.iter().cloned() {
+            if AHPForR1CS::<F>::LC_WITH_ZERO_EVAL.contains(&poly_label.as_ref()) {
+                evaluations.insert((poly_label, point), F::zero());
+            } else {
+                evaluation_labels.push((poly_label, point));
+            }
+        }
+        evaluation_labels.sort_by(|a, b| a.0.cmp(&b.0));
+        for (q, eval) in evaluation_labels.into_iter().zip(&proof.evaluations) {
+            evaluations.insert(q, *eval);
+        }
+
+        let lc_s = AHPForR1CS::construct_linear_combinations(
+            &public_input,
+            &evaluations,
+            &verifier_state,
+        )?;
+
+        let evaluations_are_correct = PC::check_combinations(
+            &index_vk.verifier_key,
+            &lc_s,
+            &commitments,
+            &query_set,
+            &evaluations,
+            &proof.pc_proof,
+            &mut opening_challenge,
+            rng,
+        )
+        .map_err(Error::from_pc_err)?;
+
+        if !evaluations_are_correct {
+            eprintln!("PC::Check failed");
+        }
+        end_timer!(verifier_time, || format!(
+            " PC::Check for AHP Verifier linear equations: {}",
+            evaluations_are_correct
+        ));
+        Ok(evaluations_are_correct)
+    }
+}
+
+impl<
+        F: PrimeField + Absorb,
+        PC: PolynomialCommitment<F, DensePolynomial<F>, S>,
+        S: CryptographicSponge + Default + RngCore,
+    > Marlin<F, PC, S>
+{
+    /// Create a zkSNARK asserting that the constraint system is satisfied.
+    /// Uses fast absorption of field elements into sponge
+    pub fn fast_prove<C: ConstraintSynthesizer<F>, R: RngCore + CryptographicSponge>(
```

Comment:
What is this new function?

Reply:
https://github.com/arkworks-rs/marlin/pull/94#discussion_r1350419122